File: src/main/java/me/hhitt/disasters/disaster/impl/LavaRising.java
package me.hhitt.disasters.disaster.impl;

import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.Disaster;
import me.hhitt.disasters.service.DeathMessageService;
import me.hhitt.disasters.storage.file.FileManager;
import me.hhitt.disasters.util.Notify;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LavaRising implements Disaster {

    private final List<Arena> arenas = new ArrayList<>();
    private final Map<Arena, Integer> currentY = new HashMap<>();

    @Override
    public void start(Arena arena) {
        arenas.add(arena);
        int minY = Math.min(arena.getCorner1().getBlockY(), arena.getCorner2().getBlockY());
        currentY.put(arena, minY);
        Notify.disaster(arena, "lava-rising");
    }

    @Override
    public void pulse(int time) {
        ConfigurationSection config = FileManager.get("config");
        int riseInterval = config != null ? config.getInt("disasters.per-disaster.lava-rising.rise-interval-seconds", 5) : 5;
        int maxBlocks = config != null ? config.getInt("disasters.per-disaster.lava-rising.max-blocks-per-pulse", 2500) : 2500;

        if (time % riseInterval != 0) return;
        for (Arena arena : new ArrayList<>(arenas)) {
            Integer y = currentY.get(arena);
            if (y == null) continue;
            fillLayer(arena, y, maxBlocks);
            currentY.put(arena, y + 1);

            for (Player player : new ArrayList<>(arena.getAlive())) {
                Material feet = player.getLocation().getBlock().getType();
                Material head = player.getLocation().clone().add(0.0, 1.0, 0.0).getBlock().getType();
                if (feet == Material.LAVA || head == Material.LAVA) {
                    DeathMessageService.mark(player, "lava-rising");
                    player.setFireTicks(Math.max(player.getFireTicks(), 80));
                    player.damage(3.0);
                }
            }
        }
    }

    @Override
    public void stop(Arena arena) {
        arenas.remove(arena);
        currentY.remove(arena);
    }

    private void fillLayer(Arena arena, int y, int maxBlocks) {
        World world = arena.getCorner1().getWorld();
        if (world == null) return;
        int minX = Math.min(arena.getCorner1().getBlockX(), arena.getCorner2().getBlockX());
        int maxX = Math.max(arena.getCorner1().getBlockX(), arena.getCorner2().getBlockX());
        int minZ = Math.min(arena.getCorner1().getBlockZ(), arena.getCorner2().getBlockZ());
        int maxZ = Math.max(arena.getCorner1().getBlockZ(), arena.getCorner2().getBlockZ());
        int placed = 0;
        int xRange = Math.max(maxX - minX, 1);
        int zRange = Math.max(maxZ - minZ, 1);
        int step = (int) Math.max(((long) xRange * zRange) / maxBlocks, 1L);
        int ix = 0;
        while (ix <= xRange && placed < maxBlocks) {
            int iz = 0;
            while (iz <= zRange && placed < maxBlocks) {
                org.bukkit.block.Block block = world.getBlockAt(minX + ix, y, minZ + iz);
                Material type = block.getType();
                if ((type == Material.AIR || type == Material.CAVE_AIR || type == Material.VOID_AIR)
                    && y + 1 < world.getMaxHeight()
                ) {
                    block.setType(Material.LAVA);
                    placed++;
                }
                iz += step;
            }
            ix += step;
        }
    }
}

File: src/main/kotlin/me/hhitt/disasters/disaster/impl/LavaRising.kt
package me.hhitt.disasters.disaster.impl

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.Disaster
import me.hhitt.disasters.storage.file.FileManager
import me.hhitt.disasters.util.Notify
import org.bukkit.Material
import me.hhitt.disasters.service.DeathMessageService

class LavaRising : Disaster {

    private val arenas = mutableListOf<Arena>()
    private val currentY = mutableMapOf<Arena, Int>()

    override fun start(arena: Arena) {
        arenas.add(arena)
        val minY = kotlin.math.min(arena.corner1.blockY, arena.corner2.blockY)
        currentY[arena] = minY
        Notify.disaster(arena, "lava-rising")
    }

    override fun pulse(time: Int) {
        val config = FileManager.get("config")
        val riseInterval = config?.getInt("disasters.per-disaster.lava-rising.rise-interval-seconds", 5) ?: 5
        val maxBlocks = config?.getInt("disasters.per-disaster.lava-rising.max-blocks-per-pulse", 2500) ?: 2500

        if (time % riseInterval != 0) return
        arenas.toList().forEach { arena ->
            val y = currentY[arena] ?: return@forEach
            fillLayer(arena, y, maxBlocks)
            currentY[arena] = y + 1

            arena.alive.toList().forEach { player ->
                val feet = player.location.block.type
                val head = player.location.clone().add(0.0, 1.0, 0.0).block.type
                if (feet == Material.LAVA || head == Material.LAVA) {
                    DeathMessageService.mark(player, "lava-rising")
                    player.fireTicks = kotlin.math.max(player.fireTicks, 80)
                    player.damage(3.0)
                }
            }
        }
    }

    override fun stop(arena: Arena) {
        arenas.remove(arena)
        currentY.remove(arena)
    }

    private fun fillLayer(arena: Arena, y: Int, maxBlocks: Int) {
        val world = arena.corner1.world ?: return
        val minX = kotlin.math.min(arena.corner1.blockX, arena.corner2.blockX)
        val maxX = kotlin.math.max(arena.corner1.blockX, arena.corner2.blockX)
        val minZ = kotlin.math.min(arena.corner1.blockZ, arena.corner2.blockZ)
        val maxZ = kotlin.math.max(arena.corner1.blockZ, arena.corner2.blockZ)
        var placed = 0
        val xRange = (maxX - minX).coerceAtLeast(1)
        val zRange = (maxZ - minZ).coerceAtLeast(1)
        val step = (((xRange.toLong() * zRange) / maxBlocks.toLong()).coerceAtLeast(1L)).toInt()
        var ix = 0
        while (ix <= xRange && placed < maxBlocks) {
            var iz = 0
            while (iz <= zRange && placed < maxBlocks) {
                val block = world.getBlockAt(minX + ix, y, minZ + iz)
                val type = block.type
                if ((type.isAir || type == Material.CAVE_AIR || type == Material.VOID_AIR)
                    && y + 1 < world.maxHeight
                ) {
                    block.setType(Material.LAVA)
                    placed++
                }
                iz += step
            }
            ix += step
        }
    }
}
